File: mailblastr-java/src/test/java/com/mailblastr/tests/DomainsTest.java
package com.mailblastr.tests;

import com.mailblastr.Mailblastr;
import com.mailblastr.requests.ClaimDomainRequest;
import com.mailblastr.requests.CreateDomainRequest;
import com.mailblastr.requests.UpdateDomainRequest;

/** Domains incl. claim flow and one-click DNS applies. */
public final class DomainsTest {
    public static void run() {
        Check.suite("DomainsTest");
        StubTransport t = new StubTransport();
        Mailblastr mb = new Mailblastr("mb_test_key", "https://api.test", t);
        t.respond(200, "{\"object\":\"domain\",\"id\":\"dom_1\",\"status\":\"pending\"}");

        // --- create with capabilities ---
        mb.domains().create(CreateDomainRequest.builder()
                .name("example.com")
                .region("us-east-1")
                .clickTracking(true)
                .receiving(true)
                .build());
        Check.eq("domain create url", "https://api.test/domains", t.lastUrl);
        Check.eq("domain create body",
                "{\"name\":\"example.com\",\"region\":\"us-east-1\",\"click_tracking\":true,"
                        + "\"capabilities\":{\"receiving\":\"enabled\"}}",
                t.lastBody);

        // --- verify / update ---
        mb.domains().verify("dom_1");
        Check.eq("verify url", "https://api.test/domains/dom_1/verify", t.lastUrl);
        Check.eq("verify method", "POST", t.lastMethod);
        mb.domains().update("dom_1", UpdateDomainRequest.builder().tls("enforced").build());
        Check.eq("update method", "PATCH", t.lastMethod);
        Check.eq("update body", "{\"tls\":\"enforced\"}", t.lastBody);

        // --- claim flow ---
        t.respond(200, "{\"object\":\"domain_claim\",\"id\":\"clm_1\"}");
        mb.domains().claim(ClaimDomainRequest.builder().name("example.com").build());
        Check.eq("claim url", "https://api.test/domains/claim", t.lastUrl);
        Check.eq("claim body", "{\"name\":\"example.com\"}", t.lastBody);
        mb.domains().getClaim("dom_1");
        Check.eq("getClaim url", "https://api.test/domains/dom_1/claim", t.lastUrl);
        Check.eq("getClaim method", "GET", t.lastMethod);
        mb.domains().verifyClaim("dom_1");
        Check.eq("verifyClaim url", "https://api.test/domains/dom_1/claim/verify", t.lastUrl);

        // --- DNS detection + one-click applies ---
        t.respond(200, "{}");
        mb.domains().detectDns("dom_1");
        Check.eq("detectDns url", "https://api.test/domains/dom_1/dns/detect", t.lastUrl);
        mb.domains().applyCloudflareDns("dom_1", "cf_token");
        Check.eq("cloudflare url", "https://api.test/domains/dom_1/dns/cloudflare", t.lastUrl);
        Check.eq("cloudflare body", "{\"token\":\"cf_token\"}", t.lastBody);
        mb.domains().applyGoDaddyDns("dom_1", "gd_key", "gd_secret");
        Check.eq("godaddy body", "{\"key\":\"gd_key\",\"secret\":\"gd_secret\"}", t.lastBody);
        mb.domains().applyNamecheapDns("dom_1", "ncuser", "nckey", "ncname");
        Check.eq("namecheap body",
                "{\"apiUser\":\"ncuser\",\"apiKey\":\"nckey\",\"userName\":\"ncname\"}", t.lastBody);
        mb.domains().applyNamecheapDns("dom_1", "ncuser", "nckey");
        Check.eq("namecheap body without userName",
                "{\"apiUser\":\"ncuser\",\"apiKey\":\"nckey\"}", t.lastBody);

        // --- remove ---
        mb.domains().remove("dom_1");
        Check.eq("remove method", "DELETE", t.lastMethod);
        Check.eq("remove url", "https://api.test/domains/dom_1", t.lastUrl);
    }

    public static void main(String[] args) {
        run();
        Check.finish();
    }
}

File: mailblastr-php/tests/cases/DomainsTest.php
<?php

declare(strict_types=1);

[$mb, $t] = make_client();

$mb->domains->create(['name' => 'example.com', 'capabilities' => ['receiving' => 'enabled']]);
check_same('domains.create: method', 'POST', $t->last()['method']);
check_same('domains.create: path', '/domains', $t->lastPath());
check_same('domains.create: body', ['name' => 'example.com', 'capabilities' => ['receiving' => 'enabled']], $t->lastJson());
check_auth('domains.create', $t->last());

$mb->domains->get('dom_1');
check_same('domains.get: path', '/domains/dom_1', $t->lastPath());

$mb->domains->list(['limit' => 3]);
check_same('domains.list: path', '/domains?limit=3', $t->lastPath());

$mb->domains->update('dom_1', ['click_tracking' => true, 'tls' => 'enforced']);
check_same('domains.update: method', 'PATCH', $t->last()['method']);
check_same('domains.update: body', ['click_tracking' => true, 'tls' => 'enforced'], $t->lastJson());

$mb->domains->verify('dom_1');
check_same('domains.verify: method', 'POST', $t->last()['method']);
check_same('domains.verify: path', '/domains/dom_1/verify', $t->lastPath());
check_same('domains.verify: no body', null, $t->last()['body']);

// ---- claim flow ----
$mb->domains->claim(['name' => 'example.com']);
check_same('domains.claim: path', '/domains/claim', $t->lastPath());
check_same('domains.claim: body', ['name' => 'example.com'], $t->lastJson());

$mb->domains->getClaim('dom_1');
check_same('domains.getClaim: path', '/domains/dom_1/claim', $t->lastPath());

$mb->domains->verifyClaim('dom_1');
check_same('domains.verifyClaim: method', 'POST', $t->last()['method']);
check_same('domains.verifyClaim: path', '/domains/dom_1/claim/verify', $t->lastPath());

// ---- one-click DNS ----
$mb->domains->detectDns('dom_1');
check_same('domains.detectDns: path', '/domains/dom_1/dns/detect', $t->lastPath());

$mb->domains->applyCloudflareDns('dom_1', ['token' => 'cf_token']);
check_same('domains.applyCloudflareDns: path', '/domains/dom_1/dns/cloudflare', $t->lastPath());
check_same('domains.applyCloudflareDns: body', ['token' => 'cf_token'], $t->lastJson());

$mb->domains->applyGoDaddyDns('dom_1', ['key' => 'k', 'secret' => 's']);
check_same('domains.applyGoDaddyDns: path', '/domains/dom_1/dns/godaddy', $t->lastPath());

$mb->domains->applyNamecheapDns('dom_1', ['apiUser' => 'u', 'apiKey' => 'k']);
check_same('domains.applyNamecheapDns: path', '/domains/dom_1/dns/namecheap', $t->lastPath());

$mb->domains->remove('dom_1');
check_same('domains.remove: method', 'DELETE', $t->last()['method']);
check_same('domains.remove: path', '/domains/dom_1', $t->lastPath());
